shellai/config.py

```diff
@@ -0,0 +1,150 @@
+    import tomllib
+else:
+    import tomli as tomllib
+
+CONFIG_DEFAULT_PATH: Path = Path("~/.config/shellai/config.toml")
+
+# tomllib does not support writting files, so we will create our own.
+CONFIG_TEMPLATE = """
+[output]
+# otherwise recording via script session will be enforced
+enforce_script = {enforce_script}
+# file with output(s) of regular commands (e.g. ls, echo, etc.)
+file = "{output_file}"
+# Keep non-empty if your file contains only output of commands (not prompt itself)
+prompt_separator = "{prompt_separator}"
+
+[history]
+enabled = {enabled}
+file = "{history_file}"
+# max number of queries in history (including responses)
+max_size = {max_size}
+
+[backend]
+endpoint = "{endpoint}"
+
+"""
+
+
+class Output(namedtuple("Output", ["enforce_script", "file", "prompt_separator"])):
+    """This class represents the [output] section of our config.toml file."""
+
+    __slots__ = ()
+
+    def __new__(
+        cls,
+        enforce_script: bool = False,
+        file: Path = Path("/tmp/shellai_output.txt"),
+        prompt_separator: str = "$",
+    ):
+        return super(Output, cls).__new__(cls, enforce_script, file, prompt_separator)
+
+
+class History(namedtuple("History", ["enabled", "file", "max_size"])):
+    """This class represents the [history] section of our config.toml file."""
+
+    __slots__ = ()
+
+    def __new__(
+        cls,
+        enabled: bool = True,
+        file: Path = Path("~/.local/share/shellai/shellai_history.json"),
+        max_size: int = 100,
+    ):
+        file = Path(file)
+        return super(History, cls).__new__(cls, enabled, file, max_size)
+
+
+class Backend(namedtuple("Backend", ["endpoint"])):
+    """This class represents the [backend] section of our config.toml file."""
+
+    __slots__ = ()
+
+    def __new__(
+        cls,
+        endpoint: str = "http://0.0.0.0:8080/v1/query/",
+    ):
+        return super(Backend, cls).__new__(cls, endpoint)
+
+
+class Config:
+    """Class that holds our configuration file representation.
+
+    .. note::
+        With this class, after being initialized, one can access their fields like:
+
+        >>> config = Config()
+        >>> config.output.enforce_script
+
+        The currently available top-level fields are:
+            * output = Match the `py:Output` class and their fields
+            * history = Match the `py:History` class and their fields
+            * backend = Match the `py:backend` class and their fields
+    """
+
+    def __init__(self, output: dict, history: dict, backend: dict) -> None:
+        self.output: Output = Output(**output)
+        self.history: History = History(**history)
+        self.backend: Backend = Backend(**backend)
+
+
+def _create_config_file(config_file: Path) -> None:
+    """Create a new configuration file with default values."""
+
+    logging.info(f"Creating new config file at {config_file.parent}")
+    config_file.parent.mkdir(mode=0o755, exist_ok=True)
```

Comment:
This is insecure in the same way as `os.makedirs(exist_ok=True)`, mentioned below.